src/Swarm/Language/Typecheck.hs

```diff
@@ -532,6 +532,7 @@ inferConst c = case c of
   Create -> [tyQ| text -> cmd unit |]
   Time -> [tyQ| cmd int |]
   Whereami -> [tyQ| cmd (int * int) |]
+  Detect -> [tyQ| text -> ((int * int) * (int * int)) -> cmd bool |]
```

Comment:
What do you think about making the type something like `text -> ((int * int) * (int * int)) -> cmd (unit + (int * int))` instead?  That is, instead of just returning a boolean, if it finds the thing you're looking for it actually tells you (the offset to) where it is.  That seems like it would be strictly more useful.

Reply:
Good idea.

Now that I think of it, eventually we could expose a spectrum of different sensory commands with various precision/constraints.  Some ideas:

```
sniff : text -> cmd int
```
Olfactory.  Physically interpreted as a "point sensor", thus cannot infer direction.  Return the **distance** to the nearest entity of a given name.  To use it, the player can move around in different directions to see whether they are getting "hot" or "cold".

```
chirp : text -> cmd dir
```
Sonar based on two "ears" will return the **direction** to the nearest entity of a given name.  To use it, the player can move in said direction until the reported direction reverses, and then progressively home in on it (or switch to another close-range sensing modality).

```
resonate : text -> ((int * int) * (int * int)) -> cmd bool
```
The original Boolean type signature.  Usage would entail a [quadtree](https://en.wikipedia.org/wiki/Quadtree) search strategy.  The physical justification is more of a stretch, but could entail something like MRI, where a strong magnetic field is applied over a given area, and a stimulating pulse is emitted, exciting the particles of the entity in question.  A non-directional signal is returned from the excited area, with a receiver being tuned to the frequency of the entity's particles.